test_runner/regress/test_remote_storage.py

```diff
@@ -777,6 +777,103 @@ def create_in_background():
         create_thread.join()
 
 
+# Regression test for a race condition where files are compactified before the upload,
+# resulting in the uploading complaining about the file not being found
+# https://github.com/neondatabase/neon/issues/4526
+@pytest.mark.parametrize("remote_storage_kind", [RemoteStorageKind.LOCAL_FS])
+def test_compaction_delete_before_upload(
+    neon_env_builder: NeonEnvBuilder,
+    remote_storage_kind: RemoteStorageKind,
+):
+    neon_env_builder.enable_remote_storage(
+        remote_storage_kind=remote_storage_kind,
+        test_name="test_compaction_delete_before_upload",
+    )
+
+    env = neon_env_builder.init_start()
+
+    # create tenant with config that will determinstically allow
+    # compaction and disables gc
+    tenant_id, timeline_id = env.neon_cli.create_tenant(
+        conf={
+            # Set a small compaction threshold
+            "compaction_threshold": "3",
+            # Disable GC
+            "gc_period": "0s",
+            # disable PITR
+            "pitr_interval": "0s",
+        }
+    )
+
+    client = env.pageserver.http_client()
+
+    endpoint = env.endpoints.create_start("main", tenant_id=tenant_id)
+
+    # Build two tables with some data inside
+    endpoint.safe_psql_many(
+        [
+            "CREATE TABLE foo (x INTEGER)",
+            "INSERT INTO foo SELECT g FROM generate_series(1, 10000) g",
+        ]
+    )
+    wait_for_last_flush_lsn(env, endpoint, tenant_id, timeline_id)
+
+    client.timeline_checkpoint(tenant_id, timeline_id)
+
+    endpoint.safe_psql_many(
+        [
+            "CREATE TABLE bar (x INTEGER)",
+            "INSERT INTO bar SELECT g FROM generate_series(1, 10000) g",
+        ]
+    )
+    wait_for_last_flush_lsn(env, endpoint, tenant_id, timeline_id)
+
+    # Now pause the compaction
+    client.configure_failpoints(("flush-frozen-before-sync", "pause"))
+
+    endpoint.safe_psql("UPDATE foo SET x = 0 WHERE x=1")
+
+    wait_for_last_flush_lsn(env, endpoint, tenant_id, timeline_id)
+
+    q: queue.Queue[Optional[PageserverApiException]] = queue.Queue()
+    barrier = threading.Barrier(2)
+
+    def checkpoint_in_background():
+        barrier.wait()
+        try:
+            client.timeline_checkpoint(tenant_id, timeline_id)
+            q.put(None)
+        except PageserverApiException as e:
+            q.put(e)
+
+    create_thread = threading.Thread(target=checkpoint_in_background)
+    create_thread.start()
+
+    try:
+        barrier.wait()
+
+        time.sleep(4)
```

Comment:
compaction_threshold +1 ?

Reply:
To fill in what I remember; it was needed so that the timeline_checkpoint has enough time to write out the delta layer and hit the failpoint. Elsewhere in tests there are examples of `wait_until` loops checking if the log has "at failpoint xxyyzz".

Reply:
I think the idea of this sleep is to ensure ordering, to have the compaction on the main thread run after the other thread's compaction has already started (but I don't have complete understanding of this). So four is just a random number, 3 works too.

The error it gives if you remove the sleep is:

```
FAILED test_runner/regress/test_remote_storage.py::test_compaction_delete_before_upload[debug-pg14-local_fs] - assert None
```